rego/rego_test.go

```diff
@@ -2105,7 +2105,47 @@ p {
 	}
 	_, ok := ndBC["http.send"]
 	if !ok {
-		t.Errorf("expected http.send cache entry")
+		t.Fatalf("expected http.send cache entry")
+	}
+}
+
+// Catches issues around iteration with ND builtins.
+func TestNDBCacheWithRuleBodyAndIteration(t *testing.T) {
+	ctx := context.Background()
+	ndBC := builtins.NDBCache{}
+	query := "data.foo.results = x"
+	_, err := New(
+		Query(query),
+		NDBuiltinCache(ndBC),
+		Module("test.rego", `package foo
+
+import future.keywords
+
+urls := [
+	"https://httpbin.org/headers",
+	"https://httpbin.org/ip",
+	"https://httpbin.org/user-agent"
+]
+
+results[response] {
+	some url in urls
+	response := http.send({
+		"method": "GET",
+		"url": url
+	})
+}`),
```

Comment:
We're likely to run into test flakes w/ this because it has to reach out to the internet. We would be better off registering a mock built-in function that sets the Nondeterministic bit.

Reply:
...or use rand.int_n with a key from a variable...? But agreed let's find a way to do this without the internet.

Reply:
Let’s keep using http.send here, although with a mock server. The impact of breaking http.send vs. rand.int_n makes it valuable, even if the bug affects both in this case.